src/realm/dictionary.cpp

```diff
@@ -508,17 +516,52 @@ std::pair<Dictionary::Iterator, bool> Dictionary::insert(Mixed key, Mixed value)
     }
 
     auto hash = key.hash();
-    ObjKey k(int64_t(hash & 0x7FFFFFFFFFFFFFFF));
+    ObjKey k(int64_t(hash & s_hash_mask));
 
     bool old_entry = false;
-    ClusterNode::State state;
-    try {
-        // We assume that we will most likely insert new values, so we try this first
+    ClusterNode::State state = m_clusters->try_get(k);
+    if (state.index == realm::npos) {
+        // key does not already exist
         state = m_clusters->insert(k, key, value);
     }
-    catch (const KeyAlreadyUsed&) {
-        state = m_clusters->get(k);
-        old_entry = true;
+    else {
+        // Check if key matches
+        Mixed existing_key = do_get_key(state);
+        if (existing_key.compare_signed(key) == 0) {
+            old_entry = true;
+        }
+        else {
+            // We have a collision
+            m_clusters->create_collision_column();
+
+            // Create a random key and link it
+            ObjKey random_key;
+            do {
+                random_key = ObjKey(int64_t(generator() & s_hash_mask));
```

Comment:
Why does this have to be actually randomly generated this way? Could it just be a counter? Could we just take the colliding hash and increment it until we reach an "open slot"?